### ISSUE TITLE
2.3 upgrade: Exceptional future ignored: std::system_error (error system:32, Broken pipe)

### ISSUE THREAD

**Issue by amoskong** [scylladb/scylla#3641]
*Installation details*
Scylla version (or git commit hash):  2.2.0-0.20180705.240b9f122 <--> 2.3.rc1-0.20180730.adad12ddc
Cluster size: 4 
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS7

* Description:
In rolling upgrade test, we will drain the node before upgrade or rollback. We see exception backtrace in other nodes.  `Exceptional future ignored: std::system_error (error system:32, Broken pipe)`

* Test job: 
http://jenkins.cloudius-systems.com:8080/job/scylla-2.3/job/2.3-test/job/scylla-2.3-rolling-upgrade-centos-test/base_version=2.2,label=gce3-qavpc,version_desc=2-3/10/artifact/latest/job.log
The job is finished, and upgrade completed, C-s workloads completed (not fail)

* Backtrace:
```
backtrace of 2.3.rc1-0.20180730.adad12ddc node:
==============================================
0x7b3018
0x585e54
0xa8e6f3
0x86c51a
0x86e7d3
0x870789
0x584756
0x584911
0x651ff7
0x6551a5
0x73d00d
/lib64/libpthread.so.0+0x7e24
/lib64/libc.so.6+0xfebac

backtrace<seastar::current_backtrace()::<lambda(seastar::frame)> > at /usr/src/debug/scylla-2.3.rc1/seastar/util/backtrace.hh:56
 (inlined by) seastar::current_backtrace() at /usr/src/debug/scylla-2.3.rc1/seastar/util/backtrace.cc:84
seastar::report_failed_future(std::__exception_ptr::exception_ptr) at /usr/src/debug/scylla-2.3.rc1/seastar/core/reactor.cc:4448
seastar::future<>::~future() at /usr/src/debug/scylla-2.3.rc1/seastar/core/future.hh:828
 (inlined by) seastar::tls::session::close() at /usr/src/debug/scylla-2.3.rc1/seastar/net/tls.cc:978
seastar::rpc::connection::stop_send_loop() at /usr/src/debug/scylla-2.3.rc1/seastar/rpc/rpc.cc:133
operator() at /usr/src/debug/scylla-2.3.rc1/seastar/rpc/rpc.cc:681
 (inlined by) apply<seastar::rpc::client::client(const seastar::rpc::logger&, void*, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::<lambda(seastar::future<>)>, seastar::future<> > at /usr/src/debug/scylla-2.3.rc1/seastar/core/future.hh:1390
operator()<seastar::future_state<> > at /usr/src/debug/scylla-2.3.rc1/seastar/core/future.hh:995
 (inlined by) run_and_dispose at /usr/src/debug/scylla-2.3.rc1/seastar/core/future.hh:413
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-2.3.rc1/seastar/core/reactor.cc:2620
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-2.3.rc1/seastar/core/reactor.cc:3032
seastar::reactor::run_some_tasks() at /opt/scylladb/include/c++/7/chrono:377
 (inlined by) seastar::reactor::run() at /usr/src/debug/scylla-2.3.rc1/seastar/core/reactor.cc:3179
operator() at /usr/src/debug/scylla-2.3.rc1/seastar/core/reactor.cc:4222
std::function<void ()>::operator()() const at /opt/scylladb/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::posix_thread::start_routine(void*) at /usr/src/debug/scylla-2.3.rc1/seastar/core/posix.cc:52
start_thread at pthread_create.c:?
__clone at ??:?

backtrace of 2.2.0-0.20180705.240b9f122 node:
============================================
0x5945d8
0x915b45
0x592ee6
0x5930a1
0x658ca7
0x65f593
0x72eaed
/lib64/libpthread.so.0+0x7e24
/lib64/libc.so.6+0xfebac

seastar::report_failed_future(std::__exception_ptr::exception_ptr) at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:4227
seastar::promise<>::abandoned() at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:1250
 (inlined by) seastar::promise<>::~promise() at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:458
 (inlined by) _ZZN7seastar6futureIJEE12then_wrappedINS1_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES1_EET0_OT_ENUlSA_E_D4Ev at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:983
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedINS2_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES2_EET0_OT_EUlSB_E_JEED4Ev at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:409
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedINS2_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES2_EET0_OT_EUlSB_E_JEED4Ev at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:409
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedINS2_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES2_EET0_OT_EUlSB_E_JEE15run_and_disposeEv at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:414
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2500
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2910
seastar::reactor::run_some_tasks() at /opt/scylladb/include/c++/7/chrono:377
 (inlined by) seastar::reactor::run() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:3057
operator() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:4002
std::function<void ()>::operator()() const at /opt/scylladb/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::posix_thread::start_routine(void*) at /usr/src/debug/scylla-2.2.0/seastar/core/posix.cc:52
start_thread at pthread_create.c:?
__clone at ??:?

====
0x5945d8
0x915b45
0x9601e3
0x592ee6
0x2c94340
0x5930a1
0x2c95319
0x658ca7
0x592ee6
0x65f593
0x5930a1

seastar::report_failed_future(std::__exception_ptr::exception_ptr) at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:4227
seastar::promise<>::abandoned() at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:1250
 (inlined by) seastar::promise<>::~promise() at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:458
 (inlined by) _ZZN7seastar6futureIJEE12then_wrappedINS1_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES1_EET0_OT_ENUlSA_E_D4Ev at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:983
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedINS2_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES2_EET0_OT_EUlSB_E_JEED4Ev at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:409
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedINS2_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES2_EET0_OT_EUlSB_E_JEED4Ev at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:409
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedINS2_12finally_bodyIZNS_3tls7session5closeEvEUlvE0_Lb0EEES2_EET0_OT_EUlSB_E_JEE15run_and_disposeEv at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:414
seastar::future<>::~future() at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:828
 (inlined by) seastar::tls::session::close() at /usr/src/debug/scylla-2.2.0/seastar/net/tls.cc:978
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2500
seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::connection::stop_send_loop() at /usr/src/debug/scylla-2.2.0/seastar/rpc/rpc.hh:226
 (inlined by) seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}::operator()(seastar::future<>) const at /usr/src/debug/scylla-2.2.0/seastar/rpc/rpc_impl.hh:1109
 (inlined by) std::enable_if<!seastar::is_future<std::result_of<seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2} (seastar::future<>&&)>::type>::value, seastar::future<> >::type seastar::do_void_futurize_apply<seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}, seastar::future<> >(std::result_of&&, seastar::future<>&&) at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:1307
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2910
seastar::future<> seastar::futurize<void>::apply<seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}, seastar::future<> >(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}&&, seastar::future<>&&) at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:1355
 (inlined by) _ZZN7seastar6futureIJEE12then_wrappedIZNS_3rpc8protocolIN4netw10serializerENS5_14messaging_verbEE6clientC4ERS8_NS3_14client_optionsENS_6socketENS_9ipv4_addrESD_EUlS1_E0_S1_EET0_OT_ENUlSH_E_clINS_12future_stateIJEEEEEDaSH_ at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:984
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedIZNS_3rpc8protocolIN4netw10serializerENS6_14messaging_verbEE6clientC4ERS9_NS4_14client_optionsENS_6socketENS_9ipv4_addrESE_EUlS2_E0_S2_EET0_OT_EUlSI_E_JEE15run_and_disposeEv at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:413
seastar::reactor::run_some_tasks() at /opt/scylladb/include/c++/7/chrono:377
 (inlined by) seastar::reactor::run() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:3057
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2500
operator() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:4002
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2910

====
0x5945d8
 0x72eaed
 0x9601e3
0x658ca7
 0x2c94340
 /lib64/libpthread.so.0+0x7e24
0x784a5f
 0x2c95319
 /lib64/libc.so.6+0xfebac
 0x592ee6
0x51ce5e
 0x5930a1
/lib64/libc.so.6+0x22444

seastar::report_failed_future(std::__exception_ptr::exception_ptr) at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:4227
std::function<void ()>::operator()() const at /opt/scylladb/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::posix_thread::start_routine(void*) at /usr/src/debug/scylla-2.2.0/seastar/core/posix.cc:52
seastar::future<>::~future() at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:828
 (inlined by) seastar::tls::session::close() at /usr/src/debug/scylla-2.2.0/seastar/net/tls.cc:978
seastar::reactor::run_some_tasks() at /opt/scylladb/include/c++/7/chrono:377
 (inlined by) seastar::reactor::run() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:3057
seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::connection::stop_send_loop() at /usr/src/debug/scylla-2.2.0/seastar/rpc/rpc.hh:226
 (inlined by) seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}::operator()(seastar::future<>) const at /usr/src/debug/scylla-2.2.0/seastar/rpc/rpc_impl.hh:1109
 (inlined by) std::enable_if<!seastar::is_future<std::result_of<seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2} (seastar::future<>&&)>::type>::value, seastar::future<> >::type seastar::do_void_futurize_apply<seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}, seastar::future<> >(std::result_of&&, seastar::future<>&&) at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:1307
start_thread at pthread_create.c:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at /usr/src/debug/scylla-2.2.0/seastar/core/app-template.cc:180
seastar::future<> seastar::futurize<void>::apply<seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}, seastar::future<> >(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>::client::client(seastar::rpc::protocol<netw::serializer, netw::messaging_verb>&, seastar::rpc::client_options, seastar::socket, seastar::ipv4_addr, seastar::ipv4_addr)::{lambda(seastar::future<>)#2}&&, seastar::future<>&&) at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:1355
 (inlined by) _ZZN7seastar6futureIJEE12then_wrappedIZNS_3rpc8protocolIN4netw10serializerENS5_14messaging_verbEE6clientC4ERS8_NS3_14client_optionsENS_6socketENS_9ipv4_addrESD_EUlS1_E0_S1_EET0_OT_ENUlSH_E_clINS_12future_stateIJEEEEEDaSH_ at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:984
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJEE12then_wrappedIZNS_3rpc8protocolIN4netw10serializerENS6_14messaging_verbEE6clientC4ERS9_NS4_14client_optionsENS_6socketENS_9ipv4_addrESE_EUlS2_E0_S2_EET0_OT_EUlSI_E_JEE15run_and_disposeEv at /usr/src/debug/scylla-2.2.0/seastar/core/future.hh:413
__clone at ??:?
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2500
main at /usr/src/debug/scylla-2.2.0/main.cc:749
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-2.2.0/seastar/core/reactor.cc:2910
__libc_start_main at ??:?
```

**Comment by avikivity** [comment]
Looks benign, but better to fix. /cc @gleb-cloudius

**Comment by gleb-cloudius** [comment]
Supposed to be fixed on master by
60892e8ce47f1201411b487e3bfecbfb0044fde6, but Calle may know better.

--
			Gleb.

**Comment by avikivity** [comment]
attn: @elcallio

**Comment by gleb-cloudius** [comment]
On Mon, Aug 06, 2018 at 01:23:16AM -0700, Calle Wilund wrote:
> Afaict this is not spuriously "leaked" future-exceptions from tls shutdown, but rpc not handling the fact that fd::shutdown_output() can throw exceptions. We are not contended on the tls socket, but the underlying socket has lost connection. tls::session::close tries to send shutdown handshake, and immediately gets failure from underlying socket -> exception. 
> 
> We can either add exception handling to rpc::stop_send_loop, or declare shutdown_output noexcept, in which case the tls code needs to add exception handlers and swallow any such. 
> 
fd::shutdown_output() is a simple void function. Making it noexcept make
more sense than handling exceptions everywhere it is called.

--
			Gleb.

**Comment by elcallio** [comment]
I would agree, but I leave these high level strategic decisions to the 
boss-men.


Den 2018-08-06 kl. 10:28, skrev Gleb Natapov:

**Comment by elcallio** [comment]
@amoskong , was this brach-2.3? If I read it correctly, it does not have scylladb/seastar@60892e8ce47f1201411b487e3bfecbfb0044fde6?

**Comment by slivne** [comment]
@avikivity - please backport

**Comment by slivne** [comment]
fixed e4d6577ef2cd5593e9ae3e15c98505f9c6f897dc
